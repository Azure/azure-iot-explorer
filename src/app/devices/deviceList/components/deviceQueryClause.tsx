/***********************************************************
 * Copyright (c) Microsoft Corporation. All rights reserved.
 * Licensed under the MIT License
 **********************************************************/
import * as React from 'react';
import { Dropdown, IDropdown, IDropdownOption } from 'office-ui-fabric-react/lib/Dropdown';
import { IconButton } from 'office-ui-fabric-react/lib/Button';
import { TextField } from 'office-ui-fabric-react/lib/TextField';
import { QueryClause, ParameterType, OperationType, DeviceCapability, DeviceStatus } from '../../../api/models/deviceQuery';
import { ResourceKeys } from '../../../../localization/resourceKeys';
import { LocalizationContextConsumer, LocalizationContextInterface } from '../../../shared/contexts/localizationContext';
import '../../../css/_deviceListQuery.scss';

export interface DeviceQueryClauseProps extends QueryClause {
    index: number;
}
export interface DeviceQueryClauseActions {
    removeClause: (index: number) => void;
    setClause: (index: number, clause: QueryClause & { isError: boolean }) => void;
}

export default class DeviceQueryClause extends React.PureComponent<DeviceQueryClauseProps & DeviceQueryClauseActions> {
    private readonly parameterTypeRef = React.createRef<IDropdown>();

    private readonly isError = (props: QueryClause) => {
        return (!props.parameterType ||
            (!props.value || '' === props.value));
    }

    private readonly remove = () => {
        this.props.removeClause(this.props.index);
    }

    private readonly onValueChange = (event: React.FormEvent<HTMLInputElement | HTMLTextAreaElement>, value: string) => {
        this.props.setClause(
            this.props.index, {
                isError: this.isError({
                    operation: this.props.operation,
                    parameterType: this.props.parameterType,
                    value
                }),
                operation: this.props.operation,
                parameterType: this.props.parameterType,
                value
            }
        );
    }

    private readonly onValueDropdownChange = (event: React.FormEvent<HTMLDivElement>, option?: IDropdownOption) => {
        this.props.setClause(
            this.props.index, {
                isError: this.isError({
                    operation: this.props.operation,
                    parameterType: this.props.parameterType,
                    value: option.key as string
                }),
                operation: this.props.operation,
                parameterType: this.props.parameterType,
                value: option.key as string
            }
        );
    }

    private readonly onTypeChange = (event: React.FormEvent<HTMLDivElement>, option?: IDropdownOption) => {
        const parameterType = option.key as ParameterType;
        this.props.setClause(
            this.props.index, {
                isError: this.isError({
                    operation: this.props.operation,
                    parameterType,
                    value: undefined
                }),
                operation: this.props.operation,
                parameterType,
                value: undefined
            }
        );
    }

    private readonly onOperationChange = (event: React.FormEvent<HTMLDivElement>, option?: IDropdownOption) => {
        const operation = option.key as OperationType;
        this.props.setClause(
            this.props.index, {
                isError: this.isError({
                    operation,
                    parameterType: this.props.parameterType,
                    value: this.props.value
                }),
                operation,
                parameterType: this.props.parameterType,
                value: this.props.value
            }
        );
    }

    public componentDidMount() {
        // set focus to parameter type dropdown
        const node = this.parameterTypeRef.current;
        if (!!node) {
            node.focus();
        }
    }

    public render(): JSX.Element {
        const isError = this.isError(this.props);
        return (
            <LocalizationContextConsumer>
                {(context: LocalizationContextInterface) => (
                    <section
                        key={`query-${this.props.index}`}
                        className={`search-pill active${isError ? ' error' : ''}${' no-operator'}`}
                    >
                        {this.renderParameterDropdown(context)}
                        {/* {this.renderOperationDropdown(context)} */}
                        {this.renderValueInput(context)}
                        <IconButton
                            className="remove-pill"
                            data={this.props.index}
                            iconProps={{ iconName: 'Clear' }}
                            type="button"
                            onClick={this.remove}
                            title={context.t(ResourceKeys.deviceLists.query.searchPills.clause.remove.title)}
                            ariaLabel={context.t(ResourceKeys.deviceLists.query.searchPills.clause.remove.ariaLabel)}
                        />
                    </section>
                )}
            </LocalizationContextConsumer>
        );
    }

    private readonly renderParameterDropdown = (context: LocalizationContextInterface) => {
        return (
            <Dropdown
                className="parameter-type"
                options={Object.keys(ParameterType).map(
                    parameterType => ({
                        key: (ParameterType as any)[parameterType], // tslint:disable-line: no-any
                        text: context.t((ResourceKeys.deviceLists.query.searchPills.clause.parameterType.items as any)[parameterType])// tslint:disable-line: no-any
                    }))}
                onChange={this.onTypeChange}
                title={context.t(ResourceKeys.deviceLists.query.searchPills.clause.parameterType.title)}
                ariaLabel={context.t(ResourceKeys.deviceLists.query.searchPills.clause.parameterType.ariaLabel)}
                componentRef={this.parameterTypeRef}
            />
        );
    }

    private readonly renderOperationDropdown = (context: LocalizationContextInterface) => {
        return (
                <Dropdown
                    className="operation-type"
                    options={Object.keys(OperationType).map
                        (operationType => ({
                            ariaLabel: context.t((ResourceKeys.deviceLists.query.searchPills.clause.operationType.options as any)[operationType]), // tslint:disable-line: no-any
                            key: (OperationType as any)[operationType], // tslint:disable-line: no-any
                            text: (OperationType as any)[operationType] // tslint:disable-line: no-any
                        }))}
                    onChange={this.onOperationChange}
                    defaultSelectedKey={this.props.operation}
                    title={context.t(ResourceKeys.deviceLists.query.searchPills.clause.operationType.title)}
                    ariaLabel={context.t(ResourceKeys.deviceLists.query.searchPills.clause.operationType.ariaLabel)}
                />
        );
    }

    private readonly renderEdgeDropdownOptions = (context: LocalizationContextInterface) => {
        return Object.keys(DeviceCapability).map
        (deviceCapability => ({
            ariaLabel: context.t((ResourceKeys.deviceLists.query.searchPills.clause.value.deviceCapability as any)[deviceCapability]), // tslint:disable-line: no-any
            key: (DeviceCapability as any)[deviceCapability], // tslint:disable-line: no-any
            text: context.t((ResourceKeys.deviceLists.query.searchPills.clause.value.deviceCapability as any)[deviceCapability]), // tslint:disable-line: no-any
        }));
    }

    private readonly renderStatusDropdownOptions = (context: LocalizationContextInterface) => {
        return Object.keys(DeviceStatus).map
        (deviceStatus => ({
            key: (DeviceStatus as any)[deviceStatus], // tslint:disable-line: no-any
            text: context.t((ResourceKeys.deviceLists.query.searchPills.clause.value.deviceStatus as any)[deviceStatus]), // tslint:disable-line: no-any
        }));
    }

    // tslint:disable-next-line:cyclomatic-complexity
    private readonly renderValueInput =  (context: LocalizationContextInterface) => {
        switch (this.props.parameterType) {
            case ParameterType.capabilityModelId:
            case ParameterType.interfaceId:
                return (
                    <TextField
                        className="clause-value"
                        placeholder={context.t(ResourceKeys.deviceLists.query.searchPills.clause.value.placeholder)}
                        ariaLabel={context.t(ResourceKeys.deviceLists.query.searchPills.clause.value.title)}
                        required={true}
                        title={context.t(ResourceKeys.deviceLists.query.searchPills.clause.value.title)}
                        onChange={this.onValueChange}
                    />
                );
            case ParameterType.edge:
                return (
                    <Dropdown
                        className="clause-value"
                        options={this.renderEdgeDropdownOptions(context)}
                        onChange={this.onValueDropdownChange}
                        title={context.t(ResourceKeys.deviceLists.query.searchPills.clause.value.title)}
                        ariaLabel={context.t(ResourceKeys.deviceLists.query.searchPills.clause.value.title)}
                        selectedKey={this.props.value || ''}
                    />
                );
            case ParameterType.status:
                return (
                    <Dropdown
                        className="clause-value"
                        options={this.renderStatusDropdownOptions(context)}
                        onChange={this.onValueDropdownChange}
                        title={context.t(ResourceKeys.deviceLists.query.searchPills.clause.value.title)}
                        ariaLabel={context.t(ResourceKeys.deviceLists.query.searchPills.clause.value.title)}
                        selectedKey={this.props.value || ''}
                    />
                );
            default: return <></>;
        }
    }
}
