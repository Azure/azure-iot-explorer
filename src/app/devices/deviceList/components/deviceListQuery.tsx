/***********************************************************
 * Copyright (c) Microsoft Corporation. All rights reserved.
 * Licensed under the MIT License
 **********************************************************/
import * as React from 'react';
import { useTranslation } from 'react-i18next';
import { TextField, ActionButton, IconButton, PrimaryButton } from '@fluentui/react';
import { DeviceQueryClause } from './deviceQueryClause';
import { DeviceQuery, QueryClause } from '../../../api/models/deviceQuery';
import { ResourceKeys } from '../../../../localization/resourceKeys';
import '../../../css/_deviceListQuery.scss';

export interface DeviceListQueryProps {
    refresh: number;
    setQueryAndExecute: (query: DeviceQuery, executeQuery?: boolean) => void;
}

export const DeviceListQuery: React.FC<DeviceListQueryProps> = (props: DeviceListQueryProps) => {
    const { t } = useTranslation();

    const { refresh, setQueryAndExecute } = props;
    const [ clauses, setClauses ] = React.useState([]);
    const [ deviceId, setDeviceId ] = React.useState('');

    React.useEffect(() => {
        if (refresh > 0) {
            setClauses(clauses);
            setDeviceId('');
        }

    },              [refresh]);

    const removePill = (index: number) => {
        const newClauses = [...clauses];
        newClauses.splice(index, 1);
        setClauses(newClauses);
    };

    const setClause = (index: number, clause: QueryClause) => {
        const newClauses = [...clauses];
        newClauses[index] = clause;
        setClauses(newClauses);
    };

    const onAddClause = () => {
        setClauses( [...clauses, { isError: true }]);
        setDeviceId('');
    };

    const onDeviceIdKeyUp = (arg: {key: string}) => {
        if (arg && 'Enter' === arg.key) {
            setQueryFromSearchBoxAndExecute();
        }
    };

    const onSetDeviceId = (event: React.FormEvent<HTMLInputElement | HTMLTextAreaElement>, newValue?: string) => {
        setClauses([]);
        setDeviceId(newValue);
    };

    const setQueryFromSearchBoxAndExecute = () => {
        setQueryAndExecute(
            {
                clauses: [],
                continuationTokens: [],
                currentPageIndex: 0,
                deviceId
            }
        );
    };

    const setQueryFromQueryPillsAndExecute = () => {
        setQueryAndExecute(
            {
                clauses,
                continuationTokens: [],
                currentPageIndex: 0,
                deviceId: ''
            }
        );
    };

    const renderIdSearch = () => {
        return (
            <div className="deviceId-search">
                <TextField
                    borderless={true}
                    placeholder={t(ResourceKeys.deviceLists.query.deviceId.placeholder)}
                    ariaLabel={t(ResourceKeys.deviceLists.query.deviceId.ariaLabel)}
                    className="search-box"
                    iconProps={{ iconName: 'Search' }}
                    role="searchbox"
                    onKeyUp={onDeviceIdKeyUp}
                    value={deviceId}
                    onChange={onSetDeviceId}
                />
                <IconButton
                    className="search-button"
                    iconProps={{ iconName: 'Forward' }}
                    type="submit"
                    onClick={setQueryFromSearchBoxAndExecute}
                    title={t(ResourceKeys.deviceLists.query.deviceId.searchButton.title)}
                    ariaLabel={t(ResourceKeys.deviceLists.query.deviceId.searchButton.ariaLabel)}
                    disabled={!!clauses && clauses.length > 0}
                />
            </div>
        );
    };

    const renderClauses = () => {
        return (
            <div className="clauses">
                {(clauses || []).map((clause, index) => <DeviceQueryClause
                    key={`query_clause:${index}`}
                    index={index}
                    operation={clause.operation}
                    parameterType={clause.parameterType}
                    removeClause={removePill}
                    setClause={setClause}
                    value={clause.value}
                />)}
                {clauses && clauses.length > 0 && <PrimaryButton
                    className="search-pill"
                    onClick={setQueryFromQueryPillsAndExecute}
                    ariaLabel={t(ResourceKeys.deviceLists.query.searchPills.search.ariaLabel)}
                >
                    {t(ResourceKeys.deviceLists.query.searchPills.search.text)}
                </PrimaryButton>}
                <ActionButton
                    className="search-pill"
                    iconProps={{ iconName: 'Filter' }}
                    type="button"
                    onClick={onAddClause}
                    ariaLabel={t(ResourceKeys.deviceLists.query.searchPills.add.ariaLabel)}
                >
                    {t(ResourceKeys.deviceLists.query.searchPills.add.text)}
                </ActionButton>
            </div>
        );
    };

    return (
        <section role="search" className="deviceList-query">
            {renderIdSearch()}
            {renderClauses()}
        </section>
    );
};
