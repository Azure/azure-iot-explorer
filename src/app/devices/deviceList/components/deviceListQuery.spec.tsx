/***********************************************************
 * Copyright (c) Microsoft Corporation. All rights reserved.
 * Licensed under the MIT License
 **********************************************************/
import 'jest';
import * as React from 'react';
import { act } from 'react-dom/test-utils';
import { shallow, mount } from 'enzyme';
import { ActionButton, IconButton, PrimaryButton, TextField } from '@fluentui/react';
import { DeviceListQuery, DeviceListQueryProps } from './deviceListQuery';
import { DeviceQueryClause } from './deviceQueryClause';

describe('DeviceListQuery', () => {
    const mockSetQueryAndExecute = jest.fn();
    const deviceListQueryProps: DeviceListQueryProps = {
        refresh: 0,
        setQueryAndExecute: mockSetQueryAndExecute
    };

    const getComponent = () => {
        return (
            <DeviceListQuery
                {...deviceListQueryProps}
            />
        );
    };

    it('matches snapshot', () => {
        expect(shallow(getComponent())).toMatchSnapshot();
    });

    it('sets device id', () => {
        const wrapper = mount(getComponent());
        const textField = wrapper.find(TextField).first();
        act(() => textField.props().onChange?.(undefined as any, 'testDevice'));
        wrapper.update();

        expect(wrapper.find(TextField).props().value).toEqual('testDevice');
        wrapper.find(IconButton).first().props().onClick(null);
        expect(mockSetQueryAndExecute).toBeCalled();
    });

    it('adds query pills and execute query', () => {
        const wrapper = mount(getComponent());
        act(() => wrapper.find(ActionButton).first().props().onClick(null));
        wrapper.update();

        expect(wrapper.find(DeviceQueryClause)).toHaveLength(1);
        expect(wrapper.find(PrimaryButton)).toBeDefined();

        act(() => wrapper.find(PrimaryButton).first().props().onClick(null));
        expect(mockSetQueryAndExecute).toBeCalled();
    });
});
