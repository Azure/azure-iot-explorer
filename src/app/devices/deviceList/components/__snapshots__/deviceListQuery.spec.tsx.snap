// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DeviceListQuery matches snapshot 1`] = `
<section
  className="deviceList-query"
  role="search"
>
  <div
    className="deviceId-search"
  >
    <StyledTextFieldBase
      ariaLabel="deviceLists.query.deviceId.ariaLabel"
      borderless={true}
      className="search-box"
      iconProps={
        Object {
          "iconName": "Search",
        }
      }
      onChange={[Function]}
      onKeyUp={[Function]}
      placeholder="deviceLists.query.deviceId.placeholder"
      role="searchbox"
      value=""
    />
    <CustomizedIconButton
      ariaLabel="deviceLists.query.deviceId.searchButton.ariaLabel"
      className="search-button"
      disabled={false}
      iconProps={
        Object {
          "iconName": "Forward",
        }
      }
      onClick={[Function]}
      title="deviceLists.query.deviceId.searchButton.title"
      type="submit"
    />
  </div>
  <div
    className="clauses"
  >
    <CustomizedActionButton
      ariaLabel="deviceLists.query.searchPills.add.ariaLabel"
      className="search-pill"
      iconProps={
        Object {
          "iconName": "Filter",
        }
      }
      onClick={[Function]}
      type="button"
    >
      deviceLists.query.searchPills.add.text
    </CustomizedActionButton>
  </div>
</section>
`;
